lasagne/layers/noise.py

```diff
@@ -80,29 +81,36 @@ def get_output_for(self, input, deterministic=False, **kwargs):
         if deterministic or self.p == 0:
             return input
         else:
-            # Using theano constant to prevent upcasting
-            one = T.constant(1)
-
-            retain_prob = one - self.p
-            if self.rescale:
-                input /= retain_prob
-
-            # use nonsymbolic shape for dropout mask if possible
-            mask_shape = self.input_shape
-            if any(s is None for s in mask_shape):
-                mask_shape = input.shape
-
-            # apply dropout, respecting shared axes
-            if self.shared_axes:
-                shared_axes = tuple(a if a >= 0 else a + input.ndim
-                                    for a in self.shared_axes)
-                mask_shape = tuple(1 if a in shared_axes else s
-                                   for a, s in enumerate(mask_shape))
-            mask = self._srng.binomial(mask_shape, p=retain_prob,
-                                       dtype=input.dtype)
-            if self.shared_axes:
-                bcast = tuple(bool(s == 1) for s in mask_shape)
-                mask = T.patternbroadcast(mask, bcast)
+            return self.apply_dropout(input, const=0)
+
+    def apply_dropout(self, input, const=0):
+        # Using theano constant to prevent upcasting
+        one = T.constant(1)
+
+        self.q = one - self.p
+        if self.rescale:
+            input /= self.q
```

Comment:
`self.q` should not be set here, it should become a property of the layer (i.e., not an attribute):
```python
@property
def q(self):
    return T.constant(1) - self.p
```

Then you never need to set `self.q` and can use it in `apply_dropout()` and in `get_output_for()`.